### ISSUE TITLE
[💡SUG] Clarification of "general" vs "context-aware"

### ISSUE THREAD

**Issue by deklanw** [RUCAIBox/RecBole#587]
I'm a bit confused about the distinction between "general" and "context-aware".

Many of the models which are here listed as "general" can incorporate user/item features.

NeuMF paper

> they can be customized to support a wide range of modelling of users and items, such as  context-aware  [28, 1],  content-based  [3],  and  neighbor-based [26].  Since this work focuses on the pure collaborative filtering setting,  we use only the identity of a user and an item  as  the  input  feature,  transforming it to a binarized sparse vector with one-hot encoding.  Note that with such a generic feature representation for inputs, our method can be easily adjusted to address the cold-start problem by using content features to represent users and items.

GCMC paper

> The item embedding v_i is calculated analogously with the same parameter matrix W.  In the presence of user- and item-specific side information we use separate parameter matrices for user and item embeddings.

etc.

Are the contextual models just the ones which have predefined ways of embedding features? Suppose I have the Movielens dataset and I want to use just the IDs as user embeddings, but for item embeddings I want to use the IDs and the genre tags. Or, suppose I want to embed the movie synopses with Word2Vec. Can I do that with the "general" recommendation models?

Looking at how the `ContextRecommender` class embeds features, it basically seems to do a bunch of lookup embeddings (and leaves floats as they are). How, then, would I incorporate Word2Vec of a movie synopsis with the "context-aware" models?

Relatedly, 
> double_tower_embed_input_fields
> Embed the whole feature columns in a double tower way.

What is a double tower way?

I think the docs have a typo here https://recbole.io/docs/user_guide/usage/running_different_models.html

> General recommendation models usually needs to group data by user and perform negative sampling.

> General recommendation models usually does not need to group data by user and perform negative sampling.

I'm assuming that latter one should say:
> Context-aware recommendation models usually do not need to group data by user or perform negative sampling.

Why don't context-aware models group data by user or perform negative sampling?

**Comment by batmanfly** [comment]
@deklanw This is an insightful comment. I will try to respond to you based on my research experience. Our team member will reply to you in a practical view based on our implementations. 

Actually, in industry, there are usually no such discrimination between different task settings for recommendation algorithms. In most scenarios, the recommendation goal is to make good recommendation using all kinds of available information, such as demographic features and temporal features. Most engineers develop their recommender systems mainly depending on available data and system goal. Therefore, in industrial recommender systems, we seldom see "general recommendation", "context-aware recommendation", etc. Their goal usually focus on generic types, including item recommendation, rating prediction, or CTR prediction.

The four kinds of recommendation tasks or settings are mainly from the research community, where the researchers formulate some specific task scenarios in a research-oriented way. With specified settings, the researchers can focus on the core problems in each task setting instead of considering recommender systems from scratch or in a very global picture. As you may know, in industry, a recall-and-reranking procedure is usually adopted, while the recall step has been seldom considered in the research community. There is some significant difference between academic and industrial task settings. 

For our library, "general recommendation" refers to the task setting where only the user/item identifiers are available. Such a simple task setting covers a large number of studies in the literature. As a comparison, "context-aware recommendation" refers to the task setting where other kinds of features are available. Such a task setting is originated from early works on feature-aware recommendation models, e.g., factorization machine. 

As you point out, it is difficult to make a strict discrimination  between general and context-aware recommendation.  Many context-aware models can degenerate to general models, when it has not been fed with context features; many general models can be upgraded to context-aware models once context features are injected (e.g., pretrained embeddings). 

Currently, we maintain the two types of models mainly based on the input. That is aligned to the commonly adopted setting in the literature. I think a simple way to connect the two groups of models is that we incorporate some kinds of embedding initialization or pretraining mechanism in general recommendation models. That may help decouple context modeling and general user-item interaction. But, many context-aware recommendation models (FM, xDeepFM)are developed by modeling feature interaction, which cann't be implemented with simple embedding lookup. Our team will reply on this point soon.


For the statement "Context-aware recommendation models usually do not need to group data by user or perform negative sampling.",  this is an incorrect statement in a general sense. We will revise it. In previous version, context-aware models mainly focus on the task of CTR prediction, which only computes AUC and LogLoss (indeed, many research papers follow this convention, however, this is not strictly correct if we treat context-aware recommendation as CTR prediction). In the case of CTR prediction, we usually don't need negative sampling. However, this is not a common setting for context-aware recommendation models. In our latest version, context-aware models can be evaluated with more metrics like general recommendation models, not limited to CTR prediction. This issue would be solved in the latest version.

Thanks for your suggestion.

**Comment by deklanw** [comment]
Thank you for the detailed reply. Makes sense.

> I think a simple way to connect the two groups of models is that we incorporate some kinds of embedding initialization or pretraining mechanism in general recommendation models. That may help decouple context modeling and general user-item interaction

I can imagine an API like this where you can specify one of "ids", "pretrained", or "both" for general models. LightGCN, for instance, doesn't intend to support rich feature embeddings (getting rid of the non-linearity and feature transformation matrices), but there isn't anything stopping you from using them. Worst case the performance is just worse. But, for GCMC e.g., this option would be very useful.

> But, many context-aware recommendation models (FM, xDeepFM)are developed by modeling feature interaction, which cann't be implemented with simple embedding lookup. Our team will reply on this point soon.

I think the ability to use pretrained embeddings for both context-aware and general models makes sense. (Maybe other kinds of models, I haven't looked at them)

>  In the case of CTR prediction, we usually don't need negative sampling.

I don't understand why. Can you expand on this point a bit? How are they trained?

**Comment by batmanfly** [comment]
@deklanw Let me clarity some points in detail. =)

1. For many CTR prediction datasets, you can have binary labels "clicked" and "non-clicked" for interaction records. In other words, the prediction task can been considered as a classification task with this kind of dataset. For training, you simply iterate over the datasets with both positive and negative cases; for testing, you just predict the interaction labels and compare them with the ground-truth labels. In both stages, negative sampling is often not used, since the negative cases are fixed like classification. 

2. It is true that both general and context-aware models can be enhanced with pertained models. However, a key point is how to pretrain these models. This is not a simple issue. In our library, heuristic strategies are used in NeuMF, and more complicated strategies (a BERT-like pertaining strategy) are developed in the S^3-Rec (sequential recommendation model). You may also take a read at our paper "Kun Zhou, Hui Wang, Wayne Xin Zhao, Yutao Zhu, Sirui Wang, Fuzheng Zhang, Zhongyuan Wang, Ji-Rong Wen: S3-Rec: Self-Supervised Learning for Sequential Recommendation with Mutual Information Maximization. CIKM 2020: 1893-1902". Indeed, it is not easy to implement a generic pretraing method for all kinds of models. Currently, we partially address this issue by providing the "embedding pre-loading" mechanism in our library, i.e., we can pre-train the embeddings with various methods outside our library, then save them in another file and finally load these parameters from the saved file. Our team will show an example on this issue soon. 


As future work, we will consider how to integrate the pre-training module into the library. 


Thanks again for your suggestion again.

**Comment by 2017pxy** [comment]
@deklanw Sorry for late reply. Firstly, thanks for your scrutiny about typo in our docs and we have corrected this mistake. Plus, about how to load pre-trained embedding into models in RecBole, we wrote a [doc](https://recbole.io/docs/user_guide/usage/load_pretrained_embedding.html) as a reference last week and hope it's helpful for you.
